File: mcp/page-auditor/src/lib/audit.ts
import { assertRobotsAllowed, parseRobotsTxt, isPathAllowedForRobots } from "./robots.js";
import { fetchPageHtml } from "./fetcher.js";
import { parseHtml } from "./parser.js";
import { sanitizeStringList, sanitizeText } from "./sanitize.js";
import { buildScoreBreakdown, totalScore } from "./scorer.js";
import { validateAuditUrl } from "./safety.js";
import type { AuditReport, HeadingsByLevel } from "./types.js";

function sanitizeHeadings(headings: HeadingsByLevel): HeadingsByLevel {
  return {
    h1: sanitizeStringList(headings.h1),
    h2: sanitizeStringList(headings.h2),
    h3: sanitizeStringList(headings.h3),
    h4: sanitizeStringList(headings.h4),
    h5: sanitizeStringList(headings.h5),
    h6: sanitizeStringList(headings.h6),
  };
}

export interface AuditUrlOptions {
  projectId?: string;
  auditId?: string;
}

export async function auditUrl(
  rawUrl: string,
  options: AuditUrlOptions = {}
): Promise<AuditReport> {
  const parsed = await validateAuditUrl(rawUrl);
  const url = parsed.href;

  await assertRobotsAllowed(parsed);

  const { html, finalUrl } = await fetchPageHtml(url);
  const data = parseHtml(html, finalUrl);
  const score_breakdown = buildScoreBreakdown(data);
  const seo_score = totalScore(score_breakdown);

  return {
    ...(options.auditId ? { audit_id: options.auditId } : {}),
    ...(options.projectId ? { project_id: options.projectId } : {}),
    url: finalUrl,
    fetched_at: new Date().toISOString(),
    meta_title: sanitizeText(data.meta_title),
    meta_title_length: data.meta_title_length,
    meta_description: sanitizeText(data.meta_description),
    meta_description_length: data.meta_description_length,
    headings: sanitizeHeadings(data.headings),
    word_count: data.word_count,
    links: data.links,
    images: {
      ...data.images,
      missing_alt_examples: sanitizeStringList(data.images.missing_alt_examples),
    },
    seo_score,
    score_breakdown,
    errors: [],
  };
}

/** @deprecated Use auditUrl — kept for backward compatibility */
export const auditPage = auditUrl;

File: apps/api/models/audit.py
from typing import Any

from pydantic import BaseModel, Field


class ScoreCategory(BaseModel):
    score: int
    max: int
    notes: str


class ScoreBreakdown(BaseModel):
    title: ScoreCategory
    description: ScoreCategory
    headings: ScoreCategory
    content_length: ScoreCategory
    links: ScoreCategory
    images: ScoreCategory


class HeadingsByLevel(BaseModel):
    h1: list[str] = Field(default_factory=list)
    h2: list[str] = Field(default_factory=list)
    h3: list[str] = Field(default_factory=list)
    h4: list[str] = Field(default_factory=list)
    h5: list[str] = Field(default_factory=list)
    h6: list[str] = Field(default_factory=list)


class AuditReport(BaseModel):
    url: str
    fetched_at: str
    meta_title: str | None = None
    meta_title_length: int = 0
    meta_description: str | None = None
    meta_description_length: int = 0
    headings: HeadingsByLevel = Field(default_factory=HeadingsByLevel)
    word_count: int = 0
    links: dict[str, int] = Field(default_factory=dict)
    images: dict[str, Any] = Field(default_factory=dict)
    seo_score: int = 0
    score_breakdown: ScoreBreakdown | None = None
    errors: list[str] = Field(default_factory=list)
